testing/trino-product-tests/src/main/java/io/trino/tests/product/hudi/TestHudiSparkCompatibility.java

```diff
@@ -0,0 +1,258 @@
+                    .containsOnly(expectedRows);
+            assertThat(onTrino().executeQuery("SELECT id, name, ts, dt, hh FROM hudi.default." + tableName))
+                    .containsOnly(expectedRows);
+
+            expectedRows = ImmutableList.of(row(2, "a2", 1000));
+            assertThat(onTrino().executeQuery("SELECT id, name, ts FROM hudi.default." + tableName + " WHERE dt = '2021-12-09' AND hh = '11'"))
+                    .containsOnly(expectedRows);
+        }
+        finally {
+            onHudi().executeQuery("DROP TABLE default." + tableName);
+        }
+    }
+
+    @Test(groups = {HUDI, PROFILE_SPECIFIC_TESTS})
+    public void testCopyOnWriteTableSelectAfterUpdate()
+    {
+        String tableName = "test_hudi_cow_select_after_update" + randomTableSuffix();
+        String tableDirectory = "hudi-compatibility-test-" + tableName;
+
+        createPartitionedCowTable(tableName, tableDirectory);
+
+        List<QueryAssert.Row> expectedRows = ImmutableList.of(
+                row(1, "a1"),
+                row(2, "a2"));
+
+        try {
+            assertThat(onHudi().executeQuery("SELECT id, name FROM default." + tableName))
+                    .containsOnly(expectedRows);
+            assertThat(onTrino().executeQuery("SELECT id, name FROM hudi.default." + tableName))
+                    .containsOnly(expectedRows);
+
+            onHudi().executeQuery("UPDATE default." + tableName + " SET name = 'a1_1', ts = 1001 WHERE id = 1");
+            expectedRows = ImmutableList.of(
+                    row(1, "a1_1", 1001),
+                    row(2, "a2", 1000));
+            assertThat(onHudi().executeQuery("SELECT id, name, ts FROM default." + tableName))
+                    .containsOnly(expectedRows);
+            assertThat(onTrino().executeQuery("SELECT id, name, ts FROM hudi.default." + tableName))
+                    .containsOnly(expectedRows);
+        }
+        finally {
+            onHudi().executeQuery("DROP TABLE default." + tableName);
+        }
+    }
+
+    @Test(groups = {HUDI, PROFILE_SPECIFIC_TESTS})
+    public void testMergeOnReadTableSelect()
+    {
+        String tableName = "test_hudi_mor_select_" + randomTableSuffix();
+        String tableDirectory = "hudi-compatibility-test-" + tableName;
+
+        onHudi().executeQuery(format("""
+                        CREATE TABLE default.%s (
+                          id int,
+                          name string,
+                          price int,
+                          ts bigint
+                        ) USING hudi
+                        tblproperties (
+                          type = 'mor',
+                          primaryKey = 'id',
+                          preCombineField = 'ts')
+                        LOCATION 's3://%s/%s'""",
+                tableName,
+                bucketName,
+                tableDirectory));
+
+        onHudi().executeQuery("INSERT INTO default." + tableName + " SELECT 1, 'a1', 20, 1000");
```

Comment:
Since we're testing "merge on read" it would be nice to actually have a case where a merge actually applies.

Reply:
Since, snapshot queries are not yet supported for merge on read tables, so I will just add test for read-optimized queries. We will add snapshot query test in #14786